### ISSUE TITLE
OpenSC.tokend does not detect when the card is removed

### ISSUE THREAD

**Issue by mouse07410** [OpenSC/OpenSC#612]
Symptoms:
1. Card is inserted
2. Card is PIN-unlocked
3. Keychain Access sees all the appropriate keys and certificates
4. Quit Keychain Access
5. Remove the card.
6. Click on the menubar Keychain icon - it still shows the card present and unlocked.
7. Click on the lock icon to lock the card - it claims success, and the system now thinks the card is present but locked.
8. Click on the menubar Keychain icon - the card shown as present and locked (note that at this moment there is no card physically present):
![screen shot 2015-11-24 at 22 38 24](https://cloud.githubusercontent.com/assets/5923577/11387782/4ce9d050-92fc-11e5-864d-30fc58d90303.png)

The log (debug=2) is inconclusive (to me):
```
$ tail /tmp/opensc-debug.log
0x11014f000 21:18:02.4294967852 [opensc-pkcs11] reader-pcsc.c:389:pcsc_detect_card_presence: returning with: 5
0x11014f000 21:18:03.7017557596194210350 [opensc-pkcs11] reader-pcsc.c:1239:pcsc_detect_readers: returning with: 0 (Success)
0x11014f000 21:18:03.140733193388591 [opensc-pkcs11] reader-pcsc.c:313:refresh_attributes: returning with: 0 (Success)
0x11014f000 21:18:03.4294967855 [opensc-pkcs11] reader-pcsc.c:389:pcsc_detect_card_presence: returning with: 5
0x11014f000 21:18:03.317106025398832 [opensc-pkcs11] reader-pcsc.c:313:refresh_attributes: returning with: 0 (Success)
0x11014f000 21:18:03.4294967856 [opensc-pkcs11] reader-pcsc.c:389:pcsc_detect_card_presence: returning with: 5
0x11014f000 21:18:03.4294967857 [opensc-pkcs11] reader-pcsc.c:313:refresh_attributes: returning with: 0 (Success)
0x11014f000 21:18:03.4294967857 [opensc-pkcs11] reader-pcsc.c:389:pcsc_detect_card_presence: returning with: 5
0x11014f000 21:18:03.4294967859 [opensc-pkcs11] reader-pcsc.c:313:refresh_attributes: returning with: 0 (Success)
0x11014f000 21:18:03.4294967859 [opensc-pkcs11] reader-pcsc.c:389:pcsc_detect_card_presence: returning with: 5
```

Running processes (related to tokend):
```
_tokend         14123   0.0  0.0  2471660   2976   ??  Ss   10:33PM   0:00.05 /System/Library/Frameworks/PCSC.framework/Versions/A/XPCServices/com.apple.ctkpcscd.xpc/Contents/MacOS/com.apple.ctkpcscd
_tokend         14122   0.0  0.0  2473584   2888   ??  S    10:33PM   0:00.15 /Library/Security/tokend/OpenSC.tokend/Contents/MacOS/OpenSC 5 Yubico Yubikey NEO OTP+U2F+CCID 30595cd0be7f000010e65bd0be7f00000000000022000000160000003bfc1300008131fe15597562696b65794e454f7233e10000000000000000000000
```

**Comment by LudovicRousseau** [comment]
How do you get the "menubar Keychain icon"?
I have never seen the Apple "Keychain Access" application providing that icon in the menu.

**Comment by mouse07410** [comment]
> How do you get the "menubar Keychain icon"?

By starting Keychain Access, going to *Preferences*, and selecting the checkbox *Show keychain status in menu bar*.
![screen shot 2015-11-25 at 12 09 34](https://cloud.githubusercontent.com/assets/5923577/11404031/db500fd2-936d-11e5-8ab1-118d07e3157d.png)

> I have never seen the Apple "Keychain Access" application providing that icon in the menu.

You do now. :-)

**Comment by frankmorgner** [comment]
I'm using El Capitan, but the drop down menu *never* distinguishes between the different keychain. For me, Keychain Access.app always recognizes when the card was removed.

Do you have some special setup?

**Comment by mouse07410** [comment]
> Do you have some special setup?

No, just a plain Yosemite.

> I'm using El Capitan, but the drop down menu never distinguishes between the
> different keychain.

Haven't tried El Capitan yet, so cannot comment - but I suspect it's a configuration option (probably user-settable?).

> For me, Keychain Access.app always recognizes when the card was removed.

Your setup is not very convincing, because for **your** cards OpenSC.tokend worked as-is, while for PIV it required quite a few changes. This may be yet another area where such differences in handling are exposed. There is a possibility that EL Capitan fixed some Yosemite problem wrt. PC/SC, but I rather doubt...

**Comment by frankmorgner** [comment]
So you think the menu item "lock keychain 'XXX'" does not show up for me, because I used a different card?

**Comment by martinpaljak** [comment]
FYI: With all versions of OSX up to 10.9 (maybe also 10.10, don't recall right away) I've seen cases where the USB device, if removed, is still visible from OSX PC/SC (and/or USB calls, can't recall the last time I debugged it) ( with OK 1021 and/or SCM 3500) so I've given up on flawless smart card usage on OSX. But maybe it gets better with 10.11 :)

**Comment by mouse07410** [comment]
> So you think the menu item "lock keychain 'XXX'" does not show up for me,
> because I used a different card?

I don't know what to think. However consider the following:
* OS X supports multiple keychains.
* It is not necessary and common-sense to be able to lock and unlock an individual keychain without affecting other keychains.
* Smartcards (tokens) represent separate keychains, subject to locking and unlocking.

In view of that I find it extremely unlikely that El Capitan did away with the above.  

Plus, observations of @martinpaljak seem to confirm what I reported.

**Comment by frankmorgner** [comment]
Tested again with Yubikey Neo and PIV applet. There is still no menu item specific to the token.

So on the one hand this "works for me" and on the other hand it may also be Apple's fault before 10.11. Additionally the problem you reported looks like a convenience thing, that does not limit OpenSC's functionality.

You may re-open this issue if you see a way for fixing this on 10.10 and before.

**Comment by mouse07410** [comment]
> Tested again with Yubikey Neo and PIV applet. 

Thanks. But I must add that this problem is fairly sporadic, and usually occurs when an application quits before the token is removed. For me MS Outlook is the typical culprit. A few times in a row I quit it, then removed the token, and the system thought the token was still inserted.

> There is still no menu item specific to the token.

Does it mean that you cannot lock the NEO once you supplied your PIN?  I find it strange, to say the least. Did you check the "Preferences" for Keychain Access? I have no El Capitan access (yet), but it just doesn't sound right.

> ...it may also be Apple's fault before 10.11...

Yes, it is possible.

> ...this "works for me"...

This is a weak argument - it worked for you with your CardOS card before any of the fixes we (and you!) created were applied. But I see your point - you're trying with NEO now, same card and applet.

I for sure won't find a way to fix it - don't know enough about PC/SC and OpenSC internals. I *think* the problem could be in the interface between the two, where OpenSC misses the opportunity to check with the low-level whether the card is still present; and/or the signaling mechanism that's supposed to inform the upper layers of the token removal for some reason does not do its job. Maybe a work-around is needed, like adding polling even though it appears unnecessary...

**Comment by dengert** [comment]
Sounds similar to the C_GetSessionInfo problem of not querying the card via PCSC.

Best I can tell, PCSC only informs the caller of a removed or reset card when the caller next tries to use the card.
One of these could get missed. because of the way tokend calls sc_pkcs15_* routines. Tokend may not be terminating
its connection to PCSC.

I other discussions, we have problems with PKCS#11 applications for these same reasons when multiple
applications are using the card.


On 11/30/2015 10:48 AM, Mouse wrote:
>     Tested again with Yubikey Neo and PIV applet.
>
> Thanks. But I must add that this problem is fairly sporadic, and usually occurs when an application quits before the token is removed. For me MS Outlook is the typical culprit. A few times in a row I
> quit it, then removed the token, and the system thought the token was still inserted.
>
>     There is still no menu item specific to the token.
>
> Does it mean that you cannot lock the NEO once you supplied your PIN? I find it strange, to say the least. Did you check the "Preferences" for Keychain Access? I have no El Capitan access (yet), but
> it just doesn't sound right.
>
>     ...it may also be Apple's fault before 10.11...
>
> Yes, it is possible.
>
>     ...this "works for me"...
>
> This is a weak argument - it worked for you with your CardOS card before any of the fixes we (and you!) created were applied. But I see your point - you're trying with NEO now, same card and applet.
>
> I for sure won't find a way to fix it - don't know enough about PC/SC and OpenSC internals. I /think/ the problem could be in the interface between the two, where OpenSC misses the opportunity to
> check with the low-level whether the card is still present; and/or the signaling mechanism that's supposed to inform the upper layers of the token removal for some reason does not do its job. Maybe a
> work-around is needed, like adding polling even though it appears unnecessary...
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/612#issuecomment-160686284>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by mouse07410** [comment]
> Tokend may not be terminating its connection to PCSC.

This is exactly what I'm observing.

> Best I can tell, PCSC only informs the caller of a removed or reset card when the caller
> next tries to use the card. One of these could get missed, because of the way tokend
> calls `sc_pkcs15_*` routines.

Perhaps somebody who understand tokend better than I do, could add the appropriate call...? And we'd see if that fixes this problem for 10.10?

Also, there is no hard evidence that this problem is eliminated in 10.11 - it doesn't occur every time on 10.10 either. Maybe @frankmorgner just hasn't hit it yet, or maybe it is prone to manifest only when certain apps are involved (like MS Outlook)...
